app/routes/surveys/components/Question.js

```diff
@@ -0,0 +1,237 @@
+import Dropdown from 'app/components/Dropdown';
+import styles from './surveys.css';
+import Icon from 'app/components/Icon';
+import { Link } from 'react-router';
+import { ConfirmModalWithParent } from 'app/components/Modal/ConfirmModal';
+
+type Props = {
+  updateQuestion: Object => void,
+  deleteQuestion: number => Promise<*>,
+  question: Object,
+  nr: number
+};
+
+type State = {
+  shouldUpdate: boolean,
+  options: Array<Object>,
+  pickerOpen: boolean
+};
+
+type QuestionTypeSelectorProps = {
+  questionType: number
+};
+
+function QuestionTypeSelector({ questionType }: QuestionTypeSelectorProps) {
+  const questionTypeIntToString = {
+    '1': 'Multiple Choice',
+    '2': 'Sjekkboks',
+    '3': 'Fritekst'
+  };
+  const questionTypeIntToIcon = {
+    '1': 'radio-button-on',
+    '2': 'checkbox',
+    '3': 'more'
+  };
+  return (
+    <Link className={styles.questionTypeSelector}>
+      <span style={{ width: '170px', display: 'inline-block' }}>
+        <Icon
+          name={questionTypeIntToIcon[String(questionType)]}
+          style={{ marginRight: '15px' }}
+        />
+        {questionTypeIntToString[String(questionType)]}
+      </span>
+      <Icon name="arrow-down" size={24} className={styles.typeIcon} />
+    </Link>
+  );
+}
+
+type DropdownItemsProps = {
+  setQuestionType: number => void
+};
+
+function DropdownItems({ setQuestionType }: DropdownItemsProps) {
+  return (
+    <Dropdown.List>
+      <Dropdown.ListItem>
+        <Link onClick={() => setQuestionType(1)}>
+          <strong>Multiple Choice</strong>
+          <Icon name="radio-button-on" size={24} />
+        </Link>
+      </Dropdown.ListItem>
+      <Dropdown.ListItem>
+        <Link onClick={() => setQuestionType(2)}>
+          Sjekkboks
+          <Icon name="checkbox" size={24} />
+        </Link>
+      </Dropdown.ListItem>
+      <Dropdown.ListItem>
+        <Link onClick={() => setQuestionType(3)}>
+          Fritekst
+          <Icon name="more" size={24} />
+        </Link>
+      </Dropdown.ListItem>
+    </Dropdown.List>
+  );
+}
+
+class Question extends Component<Props, State> {
+  state = {
+    shouldUpdate: false,
+    options: [{ nr: 0, optionText: '' }],
+    pickerOpen: false
+  };
+
+  componentDidMount() {
+    const options = this.props.question.options
+      .map((option, i) => ({ ...option, nr: i }))
+      .concat({
+        nr: this.props.question.options.length,
+        optionText: ''
+      });
+    this.setState({ options });
+  }
+
+  componentDidUpdate() {
+    if (this.state.shouldUpdate) {
+      this.props.updateQuestion({
+        ...this.props.question,
+        options: this.state.options
+      });
+      this.setState({ shouldUpdate: false });
+    }
+  }
+
+  updateOptions = (option: Object) => {
+    const options = this.state.options.slice();
+
+    const oldOption = options.find(o => o.nr === option.nr);
+    const changedOptionIndex = options.indexOf(oldOption || {});
+
+    if (changedOptionIndex !== -1) {
```

Comment:
this logic is pretty confusing - maybe it could be written in a better way? wanna try explaining what the function's purpose is in english? (maybe put that in a comment above `updateOptions` as well?)

Reply:
Yeah I figured it might be a bit difficult to understand the functionality from reading the code, so I added a quick comment, but I agree it's not the most clarifying. 

The basic functionality is simply to take an object as parameter and figure out if its a brand new option or an updated version of an old one, then add or change that option in state, respectively. Then, in addition, I implemeted a (hopefully) smoother version of having a "Add another option" button. If the incoming parameter a new option and this change is the very first change to that option, that means the user just started typing something in the last empty option seen in part 1 and 2 [here](https://user-images.githubusercontent.com/14221386/35241320-60d25bd6-ffb6-11e7-8d05-7e0f3dae2a6f.png). If so, that means we need a new blank option, because that's the only way to add more options.

Was that a better explanation?